<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>GeneKeepR Tutorial • nprcgenekeepr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="GeneKeepR Tutorial">
<meta property="og:description" content="nprcgenekeepr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">nprcgenekeepr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ColonyManagerTutorial.html">GeneKeepR: A Colony Manager's Tutorial</a>
    </li>
    <li>
      <a href="../articles/a2interactive.html">Interactive Use of nprcgenekeepr</a>
    </li>
    <li>
      <a href="../articles/a3manual.html">Genetic Management Tools Manual</a>
    </li>
    <li>
      <a href="../articles/shiny_app_use.html">GeneKeepR Tutorial</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rmsharp/nprcgenekeepr/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="shiny_app_use_files/header-attrs-2.7/header-attrs.js"></script><script src="shiny_app_use_files/kePrint-0.0.1/kePrint.js"></script><link href="shiny_app_use_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>GeneKeepR Tutorial</h1>
            <h3 class="subtitle">A Shiny Application for the Genetic Management of Colonies</h3>
                        <h4 class="author">R. Mark Sharp</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/rmsharp/nprcgenekeepr/blob/master/vignettes/shiny_app_use.Rmd"><code>vignettes/shiny_app_use.Rmd</code></a></small>
      <div class="hidden name"><code>shiny_app_use.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>This tutorial demonstrates the major functions used within the Shiny application provided by the <strong>nprcgenekeepr</strong> package. This is a brief tutorial that illustrates a typical workflow and does not explore all possible workflows.</p>
<p>Please provide any comments, questions, or bug reports through the GitHub issue tracker at <a href="https://github.com/rmsharp/nprcgenekeepr/issues">https://github.com/rmsharp/nprcgenekeepr/issues</a>.</p>
</div>
<div id="installation-and-help" class="section level2">
<h2 class="hasAnchor">
<a href="#installation-and-help" class="anchor"></a>Installation and Help</h2>
<p>To get the most recent version you can install <strong>nprcgenekeepr</strong> from GitHub with the following code.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"rmsharp/nprcgenekeepr"</span><span class="op">)</span></code></pre></div>
<p>All missing package dependencies should be automatically installed.</p>
</div>
<div id="online-documentation" class="section level2">
<h2 class="hasAnchor">
<a href="#online-documentation" class="anchor"></a>Online Documentation</h2>
<p>You can find the complete online documentation at <a href="https://rmsharp.github.io/nprcgenekeepr/">https://rmsharp.github.io/nprcgenekeepr/</a>.</p>
<p>At the top of the page are three menus to the right of the <code>Home</code> icon: <code>Reference</code>, <code>Articles</code>, and <code>Changelog</code>.</p>
<p>The <code>Reference</code> menu at the top of the page brings up the list of documentation for <code>Data objects</code>, <code>Major Features and Functions</code>, <code>Primary interactive functions</code> and <code>All exposed functions</code>.</p>
<p>The <code>Articles</code> menu brings up the list of vignettes, which are, except for <code>Development Plans</code>, tutorials for using the package.</p>
<p>The <code>Changelog</code> brings up a copy of the NEWS file of the package, which records the major changes made for each version.</p>
<p>This document is the <code>Application for Genetic Management of Colonies</code> vignette.</p>
</div>
<div id="running-shiny-application" class="section level2">
<h2 class="hasAnchor">
<a href="#running-shiny-application" class="anchor"></a>Running Shiny Application</h2>
<p>The toolset available within nprcgenekeepr can be used inside standard R scripts. However, it was originally designed to be used within a Shiny application that can be started with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rmsharp.github.io/nprcgenekeepr/">nprcgenekeepr</a></span><span class="op">)</span>
<span class="fu"><a href="../reference/runGeneKeepR.html">runGeneKeepR</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>This will result in the opening screen where you tell the application how to find the pedigree you will be using.</p>
</div>
<div id="input" class="section level2">
<h2 class="hasAnchor">
<a href="#input" class="anchor"></a>Input</h2>
<div id="pedigree-file-structure" class="section level3">
<h3 class="hasAnchor">
<a href="#pedigree-file-structure" class="anchor"></a>Pedigree File Structure</h3>
<p>Most of the screen is filled with information about formatting a text or Excel worksheet pedigree file.</p>
<p><img src="shiny_app_use_files/figure-html/eopening-screen-top-1.png" width="624"></p>
<hr>
<p>Scrolling down to the middle of the opening screen exposes a table that describes a pedigree file and further instructions.</p>
<p><img src="shiny_app_use_files/figure-html/eopening-screen-middle-1.png" width="624"></p>
<hr>
<p>Scrolling down to the bottom of the opening screen exposes more pedigree file instructions, a table that describes a genotype file and instructions regarding use of a genotype file.</p>
<p><img src="shiny_app_use_files/figure-html/eopening-screen-bottom-1.png" width="528"></p>
<hr>
<p>The following is an example of the pedigree file format.</p>
<p>Without genotypes:</p>
<p><img src="shiny_app_use_files/figure-html/eexamplePedigreeTutorial-1.png" width="528"></p>
<p>With genotypes:</p>
<p><img src="shiny_app_use_files/figure-html/eexamplePedigreeTutorial-with_alleles-1.png" width="528"></p>
<hr>
</div>
<div id="reading-in-the-pedigree" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-in-the-pedigree" class="anchor"></a>Reading in the Pedigree</h3>
<p>In this introductory tutorial, we will use an Excel file containing a hypothetical pedigree of macaques. We will work with the gray box on the left at the top of the screen.</p>
<p><img src="shiny_app_use_files/figure-html/eopening-screen-top-red-oval-1.png" width="528"></p>
<hr>
<p>A Microsoft Excel workbook with a single worksheet is the default file type; though comma (.csv), semi-colon (.txt), and tab (.txt) separated value files are all acceptable formats.</p>
<p>The <em>Example_Pedigree.xlsx</em> file we are using is from a CSV file created as shown below and then saved in an Excel format.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/makeExamplePedigreeFile.html">makeExamplePedigreeFile</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>Select the <strong>Browse</strong> button and select the pedigree file from your file system.</p>
<p><img src="shiny_app_use_files/figure-html/example-pedigree-1.png" width="240"></p>
<hr>
<p>It is important to make sure the minimum parent age is low enough for the animals in your pedigree. For our example pedigree, we are changing it from 4 years to 2 years of age since these macaques may reproduce as early as two years of age. This is shown below in three progressive images with the center image demonstrating how the hovertext provides an explanation of how this value is used.</p>
<p><img src="shiny_app_use_files/figure-html/example-pedigree-minParentAgeSequence-1.png" width="1056"></p>
<hr>
</div>
<div id="reading-in-a-pedigree-and-testing-for-errors" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-in-a-pedigree-and-testing-for-errors" class="anchor"></a>Reading in a Pedigree and Testing for Errors</h3>
<p>Selected <strong>Read and Check Pedigree</strong> will read in the file and test to see if the pedigree file has all of the columns needed and the pedigree is internally consistent.</p>
<p><img src="shiny_app_use_files/figure-html/read-and-check-pedigree-1.png" width="240"></p>
<hr>
<p>Several error types, shown below, are detected the application.</p>
<table class="table table">
<thead><tr>
<th style="text-align:left;">
Error
</th>
<th style="text-align:left;">
Definition
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;font-weight: bold;color: blue !important;">
failedDatabaseConnection
</td>
<td style="text-align:left;width: 30em; ">
Database connection failed: configuration or permissions are invalid
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;color: blue !important;">
missingColumns
</td>
<td style="text-align:left;width: 30em; ">
Columns that must be within the pedigree file are missing.
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;color: blue !important;">
invalidDateRows
</td>
<td style="text-align:left;width: 30em; ">
Values, which are supposed to be dates, cannot be interpreted as a date.
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;color: blue !important;">
suspiciousParents
</td>
<td style="text-align:left;width: 30em; ">
Parents were too young on the date of birth of to have been the parent.
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;color: blue !important;">
femaleSires
</td>
<td style="text-align:left;width: 30em; ">
Individuals listed as female or hermaphroditic and as a sire.
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;color: blue !important;">
maleDams
</td>
<td style="text-align:left;width: 30em; ">
Individuals are listed as male and as a dam.
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;color: blue !important;">
sireAndDam
</td>
<td style="text-align:left;width: 30em; ">
Individuals who are listed as both a sire and a dam.
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;color: blue !important;">
duplicateIds
</td>
<td style="text-align:left;width: 30em; ">
IDs listed more than once.
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;color: blue !important;">
fatalError
</td>
<td style="text-align:left;width: 30em; ">
Fatal Errors. These are unanticipated errors that should be reported to the package maintainer
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;color: blue !important;">
changedCols
</td>
<td style="text-align:left;width: 30em; ">
Columns that have been changed to conform to internal naming conventions and what they were changed to.
</td>
</tr>
</tbody>
</table>
<hr>
</div>
</div>
<div id="pedigree-browser" class="section level2">
<h2 class="hasAnchor">
<a href="#pedigree-browser" class="anchor"></a>Pedigree Browser</h2>
<p>The <strong>Pedigree Browser</strong> tab defaults to displaying 10 rows of the pedigree at a time, but you can choose to display 10, 25, 50, or 100 rows. You can choose to display UNKNOWN IDs in the rows displayed. UNKNOWN IDs (UIDs) are used to label unknown parents of animals with one known parent. The program calculates additional columns based on the input pedigree.</p>
<p><img src="shiny_app_use_files/figure-html/pb-10-rows-display-unknown-ids-1.png" width="528"></p>
<hr>
<div id="unknown-ids" class="section level3">
<h3 class="hasAnchor">
<a href="#unknown-ids" class="anchor"></a>Unknown IDs</h3>
<p>I have placed red lines under the UNKNOWN IDs in the partial pedigree list below. UNKNOWN IDs are used to label unknown parents of animals with one known parent. (Note these are found near the end of the pedigree list,) These IDs have no meaning other than they all begin with the letter <em>U</em> and are following with a left alphanumeric string of five places.</p>
<p><img src="shiny_app_use_files/figure-html/unknown-displayed-1.png" width="480"></p>
<hr>
<p>In this example pedigree, when you deselect the <strong>Display Unknown IDs</strong> checkbox. The number of rows reduces from 3,694 to 2,322, because there were 1,372 UNKNOWN animals generated when constructing the pedigree to provide sire and dam placeholders for all animals.</p>
<p><img src="shiny_app_use_files/figure-html/no-unknown-displayed-1.png" width="480"></p>
<hr>
</div>
<div id="selecting-a-pedigree-subset-focal-animals" class="section level3">
<h3 class="hasAnchor">
<a href="#selecting-a-pedigree-subset-focal-animals" class="anchor"></a>Selecting a Pedigree Subset — Focal Animals</h3>
<p>The <strong>Pedigree Browser</strong> tab displays the full pedigree by default but allows you to select a subset of the pedigree by entering a list of animals of interest (<em>focal animals</em>).</p>
<p><img src="shiny_app_use_files/figure-html/focal-animal-text-box-1.png" width="480"></p>
<hr>
<p>You can enter in the animal IDs by typing them into the text box directly as shown below (FJS7RQ, H6T2FF, HEVL3L, I04JZV, S63QDN). Deselect the <strong>Display Unknown IDs</strong> checkbox and select the <strong>Trim pedigree based on focal animals</strong> checkbox. (See top right of image below).</p>
<p>Trimming the pedigree based on focal animals will keep only animals in the pedigree that are related to the focal animals selected.</p>
<p>Select the <strong>Update Focal Animals</strong> button to tell the application to read your list of animals, trim the pedigree based on that list, and display the trimmed pedigree below. You will end up with 54 animals in your pedigree.</p>
<p><img src="shiny_app_use_files/figure-html/pedigree-browser-5-focal-animals-small-1.png" width="768"></p>
<hr>
<p>Also, you can import a list of focal animals by selecting the <strong>Browse</strong> button under <strong>Choose CSV file with focal animals</strong>. This file can be constructed by creating a simple text file with commas between animal IDs or by placing individual animal IDs on separate lines.</p>
<p>Focal animals are the list of animals that will be used in the following analysis. In most cases, we recommend using all alive animals in the breeding population. By selecting focal animals, the number of pedigree entrees does not change, but the population membership flag will be set to <code>TRUE</code> for the focal animals, and <code>FALSE</code> for all other animals.</p>
<p><img src="shiny_app_use_files/figure-html/selection-large-focal-group-1.png" width="384"></p>
<hr>
<p>After entering your list of focal animals, you can select to trim the pedigree so that it will only include relatives of the focal animals you have selected. This will reduce the number of members within the pedigree to all animals required to connect all of the focal animals in the pedigree.</p>
<p><img src="shiny_app_use_files/figure-html/select-trim-for-focal-animals-1.png" width="576"></p>
<hr>
<p>A pedigree trimmed based on focal animals will have only the relatives of those animals remaining. In this instance there are only a total of 522 focal animals and their relatives.</p>
<p><em>Note: focal animals and their relatives will only be included in the same pedigree when the original pedigree file uploaded indicates a common ancestor for them. Otherwise, focal animals and their relatives will be sorted into separate pedigrees in the output, with each separate pedigree indicated by its own number.</em></p>
<p><img src="shiny_app_use_files/figure-html/trimmed-for-focal-animals-1.png" width="384"></p>
<hr>
<p>You can remove the animals from the list of focal animals by selecting the <strong>Clear Focal Animals</strong> checkbox and selecting the <strong>Update Focal Animals</strong> button. This will read in an empty ID list, clear the box of IDs, and bring back all of the trimmed away IDs.</p>
<p><img src="shiny_app_use_files/figure-html/cleared-of-focal-animals-1.png" width="576"></p>
<p>Deselect the <strong>Clear Focal Animals</strong> checkbox and reselect the <strong>Update Focal Animals</strong> button before continuing with the tutorial so that we will be working with the trimmed pedigree.</p>
<hr>
</div>
</div>
<div id="pedigree-age-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#pedigree-age-plot" class="anchor"></a>Pedigree Age Plot</h2>
<p>The <strong>Pedigree Age Plot</strong> tab displays a standard pyramid plot for the pedigree as selected in the <strong>Pedigree Browser</strong> tab. This is showing 332 living animals from the entire example pedigree.</p>
<p><img src="shiny_app_use_files/figure-html/age-plot-1.png" width="768"></p>
<hr>
</div>
<div id="genetic-value-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#genetic-value-analysis" class="anchor"></a>Genetic Value Analysis</h2>
<p>Selecting the <strong>Genetic Value Analysis</strong> tab and enter the number of simulations and genome uniqueness threshold desired. See the <strong>Genetic Value Analysis and Breeding Group Description</strong> tab for a breakdown of the calculation.</p>
<p>We recommend trying multiple numbers of simulations to arrive at an ideal number that produces consistent results (i.e., 1,000). Genome uniqueness values are calculated using a gene-drop simulation according to MacCluer et al. (1986) and Ballou &amp; Lacy (1995), by assigning unique alleles to all pedigree founders, and simulating their segregation throughout the pedigree according to Mendelian rules. Genome uniqueness is a measure of the probability that an animal possesses founder alleles that are present in at most x other animals (usually 0-3), and thus are rare and at risk of being lost from the population. A range of 2 to 100,000 simulations may be selected. A minimum of 1,000 simulations is recommended. A genome uniqueness threshold value between 0-3 should also be selected, as desired.</p>
<p>Select the <strong>Begin Analysis</strong> button to start the gene dropping process, which you can monitor with the progress meter in the lower right corner of the display.</p>
<p>We are not aware of a systematic study of pedigree structure with this algorithm and have not performed extensive studies with pedigrees of various structures, but 1000 iterations has seemed to provide reproducible results for out pedigrees. It must be emphasized that pedigree structure is expected to affect precision when iterations are held constant.</p>
<!--  Come up wit a definition of reproducible and see if we can run and -->
<!-- automated test to find the needed number of iterations.-->
<p><img src="shiny_app_use_files/figure-html/gva-calculating-1.png" width="768"></p>
<hr>
<p>As soon as the calculations are completed, a table showing the results of the analysis is displayed in 10 record increments. The calculations for 1000 iterations of the gene dropping algorithm took 1 minute 38 seconds with the example pedigree of 3,691 animals using a MacBook Pro (Mid 2014), 2.8 GHz Intel Core i7 with 16 GB of 1600 MHz DDR3 memory.</p>
<p>Again you can select how many rows to display at once by changing the values in the <strong>Show entries</strong> selection tool.</p>
<p><img src="shiny_app_use_files/figure-html/gva-first-high-value-1.png" width="768"></p>
<hr>
<p>Searching down the table of results in the <strong>Value Designation</strong> column you can see starting at row 268 the values change from <em>High Value</em> to <em>Low Value</em>. Though not shown here, the value of <em>Undetermined</em> in the <strong>Value Designation</strong> column means the animal did not have parentage information. Infants or very young animals without assigned parents are given an “Undetermined” designation. Founders also do not have parentage information but are high value by definition.</p>
<p><img src="shiny_app_use_files/figure-html/gva-high-and-low-value-1.png" width="528"></p>
<hr>
</div>
<div id="summary-statistics" class="section level2">
<h2 class="hasAnchor">
<a href="#summary-statistics" class="anchor"></a>Summary Statistics</h2>
<p>The <strong>Summary Statistics and Plots</strong> tab used results from the <strong>Genetic Value Analysis</strong> tab. Definitions of genome uniqueness and kinship are located in the <strong>Genetic Value Analysis and Breeding Group Description</strong> tab. Additionally, definitions of founder equivalents and founder genome equivalents are located at the bottom of the <strong>Summary Statistics and Plots</strong> tab.</p>
<p><img src="shiny_app_use_files/figure-html/ss-first-view-1.png" width="768"></p>
<hr>
<p>The <strong>Export Kinship Matrix</strong> button creates a CSV file that has a row and column for each individual in the genetic analysis plus a first row and first column each containing the IDs.</p>
<p>The first few rows of such a file are shown below.</p>
<p><img src="shiny_app_use_files/figure-html/ss-kinship-matrix-1.png" width="576"></p>
<hr>
<p>The <strong>First-Order Relationships</strong> button creates a CSV file that has the following columns defined: an unnamed column for row number, <em>id</em>, <em>parents</em>, <em>offspring</em>, <em>siblings</em>, and <em>total</em>. Counts are based off known relationships. The first few rows of such a file are shown below.</p>
<p><img src="shiny_app_use_files/figure-html/first-order-relationships-1.png" width="480"></p>
<hr>
<p>The <strong>Export Female Founders</strong> and <strong>Export Male Founders</strong> buttons creates a CSV file that has the following columns defined: an unnamed column for row number, <em>id</em>, <em>sires</em>, <em>dam</em>, <em>sex</em>, <em>gen</em>, <em>birth</em>, <em>exit</em>, <em>age</em>, <em>recordStatus</em>, <em>population</em>, and <em>pedNum</em>. The first few rows of such a file are shown below.</p>
<p><img src="shiny_app_use_files/figure-html/female-founders-1.png" width="480"></p>
<hr>
<p>The six plots provide histograms and boxplots for the kinship coefficients, the Z-scores of the kinship coefficients, and the genome uniqueness scores.</p>
<p><img src="shiny_app_use_files/figure-html/ss-trimmed-all-plots-1.png" width="768"></p>
<hr>
<p>Each of the plots can be exported as a PNG file to a directory you choose.</p>
<p><img src="shiny_app_use_files/figure-html/ss-export-mean-kinship-coef-hist-plot-1.png" width="480"></p>
<hr>
</div>
<div id="breeding-group-formation" class="section level2">
<h2 class="hasAnchor">
<a href="#breeding-group-formation" class="anchor"></a>Breeding Group Formation</h2>
<p>Selecting the <strong>Breeding Group Formation</strong> tab brings forward the screen shown below. In this screen you can form breeding groups using one of three ways based on your source of animals selected under <strong>Choose one group formation workflow:</strong>.</p>
<p>Additionally, you must specify how you want to construct the breeding groups with regard to the groups’ sex ratios. The third of the three options (<em>User specified sex ratio of breeders</em>) causes the appearance of the field where you can fill in the sex ratio (F/M) that you want to have in the formed breeding groups. The sex ratio algorithm will form a group as nearly to the selected ratio as possible given the size of the group. Limits in the availability of either sex will restrict the size of the groups formed.</p>
<p><img src="shiny_app_use_files/figure-html/breeding-group-first-view-1.png" width="672"></p>
<hr>
<p>The <strong>Make Groups</strong> button appears once you select the source of animals you are going to use. However, you probably will be making additional selections using other controls on the screen.</p>
<p>The most common source of animals will be the high-value animals found by the genetic analysis.</p>
<p>You can either type in the number of groups that you want to form or select the number of groups using the arrows on the right edge of the <strong>Number of Groups Desired</strong> field, which is outlined in blue in the image below.</p>
<p><img src="shiny_app_use_files/figure-html/breeding-group-1-1.png" width="768"></p>
<hr>
<p>There are often behavioral constraints, such as preexisting social groups, that dictate the need to have some animals maintained together. This need is readily accommodated by pre-seeding groups with those social groups. You may select the <strong>Optional: Seed Groups with Specific Animals</strong> field if you decide to place some animals together within the groups because you know them to be compatible with each other.</p>
<p>This has been done in the example below using six groups with differing numbers of seed animals. Note the selection of having animals below the minimum parent age of two being grouped with their mother.</p>
<hr>
<p><img src="shiny_app_use_files/figure-html/breeding-group-6-infants-with-dam-1.png" width="576"></p>
<hr>
<p>Each group has all of the seed animals that were assigned to it plus additional animals that could be added while satisfying the requirements imposed by the selected settings. I have indicated the seed animals for the first group with red rectangles.</p>
<p><img src="shiny_app_use_files/figure-html/breeding-group-first-group-no-kinship-seeds-indicated-1.png" width="816"></p>
<hr>
<p>Display of kinship values requires that the <strong>Include kinship in display of groups</strong> checkbox be selected prior to group formation.</p>
<p>A group of ten animals was formed in the next run after choosing to include kinship and selecting the <strong>Make Groups</strong> button.</p>
<p><img src="shiny_app_use_files/figure-html/breeding-group-6-seed-grps-grp-6-kinship-1.png" width="816"></p>
<hr>
<p>The option to select a desired sex ratio allows you to select any ratio desired. However, the ratio obtained is limited by the availability of animals that meet all criteria you have set.</p>
<p><img src="shiny_app_use_files/figure-html/breeding-group-sex-ratio-specification-1.png" width="528"></p>
<p>Selecting a sex ratio of 2.5 with 6 groups as is illustrated resulted in 5 groups of 20 with a ratio of 14:6 (2.3) and 1 group of 23 with a ratio of 16:7 (2.3).</p>
<p>Groups can be individually exported into file names and locations of your choosing. The corresponding kinship matrix for each group can also be exported.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Michael Raboin, Terry Therneau, Amanda Vinson, R. Mark Sharp, Southwest National Primate Research Center NIH grant P51 RR13986, Oregon National Primate Research Center grant P51 OD011092.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
