<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ nprcgenekeepr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">nprcgenekeepr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/ColonyManagerTutorial.html">GeneKeepR: A Colony Manager's Tutorial</a>
    </li>
    <li>
      <a href="../articles/a2interactive.html">Interactive Use of nprcgenekeepr</a>
    </li>
    <li>
      <a href="../articles/a3manual.html">Genetic Management Tools Manual</a>
    </li>
    <li>
      <a href="../articles/a4Development_Plans.html">Development Plans</a>
    </li>
    <li>
      <a href="../articles/shiny_app_use.html">GeneKeepR Tutorial</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rmsharp/nprcgenekeepr/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-data-objects" class="hasAnchor"><a href="#section-data-objects" class="anchor"></a>Data objects</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="exampleNprcgenekeeprConfig.html">exampleNprcgenekeeprConfig</a></code> </p>
        </td>
        <td><p>exampleNprcgenekeeprConfig is a loadable version of the example
configuration file <code>example_nprcgenekeepr_config</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="examplePedigree.html">examplePedigree</a></code> </p>
        </td>
        <td><p>examplePedigree is a pedigree object created by <code>qcStudbook</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="finalRpt.html">finalRpt</a></code> </p>
        </td>
        <td><p>finalRpt is a list object created from the list object <em>rpt</em> prepared
by <code>reportGV</code>. It is created inside <code>orderReport</code>. This version
is at the state just prior to calling <code>rankSubjects</code> inside
<code>orderReport</code>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="focalAnimals.html">focalAnimals</a></code> </p>
        </td>
        <td><p>focalAnimals is a dataframe with one column (_id_) containing the of animal
Ids from the __examplePedigree__ pedigree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lacy1989Ped.html">lacy1989Ped</a></code> </p>
        </td>
        <td><p>lacy1989Ped small hypothetical pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lacy1989PedAlleles.html">lacy1989PedAlleles</a></code> </p>
        </td>
        <td><p>lacy1989PedAlleles is a dataframe produced by <code>geneDrop</code> on
<code>lacy1989Ped</code> with 5000 iterations.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ped1Alleles.html">ped1Alleles</a></code> </p>
        </td>
        <td><p>ped1Alleles is a dataframe created by the geneDrop function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pedDuplicateIds.html">pedDuplicateIds</a></code> </p>
        </td>
        <td><p>pedDuplicateIds is a dataframe with 9 rows and 5 columns (ego_id, sire,
dam_id, sex, birth_date) representing a full pedigree with a duplicated
record.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pedFemaleSireMaleDam.html">pedFemaleSireMaleDam</a></code> </p>
        </td>
        <td><p>pedFemaleSireMaleDam is a dataframe with 8 rows and 5 columns (ego_id, sire,
dam_id, sex, birth_date) representing a full pedigree with the errors of
having a sire labeled as female and a dam labeled as male.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pedGood.html">pedGood</a></code> </p>
        </td>
        <td><p>pedGood is a dataframe with 8 rows and 5 columns (ego_id, sire, dam_id,
sex, birth_date) representing a full pedigree with no errors.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pedInvalidDates.html">pedInvalidDates</a></code> </p>
        </td>
        <td><p>pedInvalidDates is a dataframe with 8 rows and 5 columns (ego_id, sire,
dam_id, sex, birth_date) representing a full pedigree with values in the
<code>birth_date</code> column that are not valid dates.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pedMissingBirth.html">pedMissingBirth</a></code> </p>
        </td>
        <td><p>pedMissingBirth is a dataframe with 8 rows and 5 columns (ego_id, sire,
dam_id, sex, birth_date) representing a full pedigree with no errors.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pedOne.html">pedOne</a></code> </p>
        </td>
        <td><p>pedOne is a loadable version of a pedigree file fragment used for testing
and demonstration</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pedSameMaleIsSireAndDam.html">pedSameMaleIsSireAndDam</a></code> </p>
        </td>
        <td><p>pedSameMaleIsSireAndDam is a dataframe with 8 rows and 5 columns (ego_id,
sire, dam_id, sex, birth_date) representing a full pedigree with no errors.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pedSix.html">pedSix</a></code> </p>
        </td>
        <td><p>pedSix is a loadable version of a pedigree file fragment used for testing
and demonstration</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pedWithGenotype.html">pedWithGenotype</a></code> </p>
        </td>
        <td><p>pedWithGenotype is a dataframe produced from qcPed by adding made up
genotypes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pedWithGenotypeReport.html">pedWithGenotypeReport</a></code> </p>
        </td>
        <td><p>pedWithGenotypeReport is a list containing the output of <code>reportGV</code>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qcBreeders.html">qcBreeders</a></code> </p>
        </td>
        <td><p>qcBreeders is a list of 29 baboon IDs that are potential breeders</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qcPed.html">qcPed</a></code> </p>
        </td>
        <td><p>qcPed is a dataframe with 277 rows and 6 columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qcPedGvReport.html">qcPedGvReport</a></code> </p>
        </td>
        <td><p>qcPedGvReport is a genetic value report</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rhesusGenotypes.html">rhesusGenotypes</a></code> </p>
        </td>
        <td><p>rhesusGenotypes is a dataframe with two haplotypes per animal</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rhesusPedigree.html">rhesusPedigree</a></code> </p>
        </td>
        <td><p>rhesusPedigree is a pedigree object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smallPed.html">smallPed</a></code> </p>
        </td>
        <td><p>smallPed is a hypothetical pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smallPedTree.html">smallPedTree</a></code> </p>
        </td>
        <td><p>smallPedTree is a pedigree tree made from <code>smallPed</code></p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-major-features-and-functions" class="hasAnchor"><a href="#section-major-features-and-functions" class="anchor"></a>Major Features and Functions</h2>
          <p class="section-desc"><p>Brief codebook and Major Functions</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="nprcgenekeepr.html">nprcgenekeepr</a></code> </p>
        </td>
        <td><p>Genetic Management Functions</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-primary-interactive-functions" class="hasAnchor"><a href="#section-primary-interactive-functions" class="anchor"></a>Primary interactive functions</h2>
          <p class="section-desc"><p>Funtions designed for interactive use</p></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="addBackSecondParents.html">addBackSecondParents()</a></code> </p>
        </td>
        <td><p>Add back single parents trimmed pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addErrTxt.html">addErrTxt()</a></code> </p>
        </td>
        <td><p>Concatenates any errors from nprcgenekeeprErr into narrative form</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addGenotype.html">addGenotype()</a></code> </p>
        </td>
        <td><p>Add genotype data to pedigree file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addParents.html">addParents()</a></code> </p>
        </td>
        <td><p>Add parents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addUIds.html">addUIds()</a></code> </p>
        </td>
        <td><p>Eliminates partial parentage situations by adding unique placeholder
IDs for the unknown parent.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="agePyramidPlot.html">agePyramidPlot()</a></code> </p>
        </td>
        <td><p>Form age pyramid plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assignAlleles.html">assignAlleles()</a></code> </p>
        </td>
        <td><p>Assign parent alleles randomly</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcA.html">calcA()</a></code> </p>
        </td>
        <td><p>Calculates <code>a</code>, the number of an individual's alleles that are rare in
each simulation.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcAge.html">calcAge()</a></code> </p>
        </td>
        <td><p>Calculate animal ages.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcFE.html">calcFE()</a></code> </p>
        </td>
        <td><p>Calculates founder Equivalents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcFEFG.html">calcFEFG()</a></code> </p>
        </td>
        <td><p>Calculates Founder Equivalents and Founder Genome Equivalents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcFG.html">calcFG()</a></code> </p>
        </td>
        <td><p>Calculates Founder Genome Equivalents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcGU.html">calcGU()</a></code> </p>
        </td>
        <td><p>Calculates genome uniqueness for each ID that is part of the population.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcRetention.html">calcRetention()</a></code> </p>
        </td>
        <td><p>Calculates Allelic Retention</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="checkGenotypeFile.html">checkGenotypeFile()</a></code> </p>
        </td>
        <td><p>Check genotype file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="checkParentAge.html">checkParentAge()</a></code> </p>
        </td>
        <td><p>Check parent ages to be at least <code>minParentAge</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="correctParentSex.html">correctParentSex()</a></code> </p>
        </td>
        <td><p>Sets sex for animals listed as either a sire or dam.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="countLoops.html">countLoops()</a></code> </p>
        </td>
        <td><p>Count the number of loops in a pedigree tree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createExampleFiles.html">createExampleFiles()</a></code> </p>
        </td>
        <td><p>Creates a folder with CSV files containing example pedigrees and ID lists
used to demonstrate the package.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createPedTree.html">createPedTree()</a></code> </p>
        </td>
        <td><p>Create a pedigree tree (PedTree).</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="finalRpt.html">finalRpt</a></code> </p>
        </td>
        <td><p>finalRpt is a list object created from the list object <em>rpt</em> prepared
by <code>reportGV</code>. It is created inside <code>orderReport</code>. This version
is at the state just prior to calling <code>rankSubjects</code> inside
<code>orderReport</code>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findGeneration.html">findGeneration()</a></code> </p>
        </td>
        <td><p>Determines the generation number for each id.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findLoops.html">findLoops()</a></code> </p>
        </td>
        <td><p>Find loops in a pedigree tree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findPedigreeNumber.html">findPedigreeNumber()</a></code> </p>
        </td>
        <td><p>Determines the generation number for each id.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fixGenotypeCols.html">fixGenotypeCols()</a></code> </p>
        </td>
        <td><p>Reformat names of observed genotype columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneDrop.html">geneDrop()</a></code> </p>
        </td>
        <td><p>Gene drop simulation based on the provided pedigree information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getAncestors.html">getAncestors()</a></code> </p>
        </td>
        <td><p>Recursively create a character vector of ancestors for an individual ID.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getDatedFilename.html">getDatedFilename()</a></code> </p>
        </td>
        <td><p>Returns a character vector with an file name having the date prepended.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getDemographics.html">getDemographics()</a></code> </p>
        </td>
        <td><p>Get demographic data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getFocalAnimalPed.html">getFocalAnimalPed()</a></code> </p>
        </td>
        <td><p>Get pedigree based on list of focal animals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGenotypes.html">getGenotypes()</a></code> </p>
        </td>
        <td><p>Get genotypes from file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGVGenotype.html">getGVGenotype()</a></code> </p>
        </td>
        <td><p>Get Genetic Value Genotype data structure for reportGV function.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGVPopulation.html">getGVPopulation()</a></code> </p>
        </td>
        <td><p>Get the population of interest for the Genetic Value analysis.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getLkDirectAncestors.html">getLkDirectAncestors()</a></code> </p>
        </td>
        <td><p>Get the direct ancestors of selected animals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getLkDirectRelatives.html">getLkDirectRelatives()</a></code> </p>
        </td>
        <td><p>Get the direct ancestors of selected animals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPedigree.html">getPedigree()</a></code> </p>
        </td>
        <td><p>Get pedigree from file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPedMaxAge.html">getPedMaxAge()</a></code> </p>
        </td>
        <td><p>Get the maximum age of live animals in the pedigree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPossibleCols.html">getPossibleCols()</a></code> </p>
        </td>
        <td><p>Get possible column names for a studbook.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPyramidPlot.html">getPyramidPlot()</a></code> </p>
        </td>
        <td><p>Creates a pyramid plot of the pedigree provided.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getRequiredCols.html">getRequiredCols()</a></code> </p>
        </td>
        <td><p>Get required column names for a studbook.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getSiteInfo.html">getSiteInfo()</a></code> </p>
        </td>
        <td><p>Get site information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="kinship.html">kinship()</a></code> </p>
        </td>
        <td><p>Generates a kinship matrix.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeExamplePedigreeFile.html">makeExamplePedigreeFile()</a></code> </p>
        </td>
        <td><p>Write copy of nprcgenekeepr::examplePedigree into a file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="obfuscateDate.html">obfuscateDate()</a></code> </p>
        </td>
        <td><p>obfucateDate adds a random number of days bounded by plus and minus max delta</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="obfuscateId.html">obfuscateId()</a></code> </p>
        </td>
        <td><p>obfucateId creates a vector of ID aliases of specified length</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="obfuscatePed.html">obfuscatePed()</a></code> </p>
        </td>
        <td><p>obfuscatePed takes a pedigree object and creates aliases for all IDs and
adjusts all date within a specified amount.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="print.html">print(<i>&lt;summary.nprcgenekeeprErr&gt;</i>)</a></code> <code><a href="print.html">print(<i>&lt;summary.nprcgenekeeprGV&gt;</i>)</a></code> </p>
        </td>
        <td><p>print.summary.nprcgenekeepr print.summary.nprcgenekeeprGV</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qcStudbook.html">qcStudbook()</a></code> </p>
        </td>
        <td><p>Quality Control for the Studbook or pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rankSubjects.html">rankSubjects()</a></code> </p>
        </td>
        <td><p>Ranks animals based on genetic value.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="kinMatrix2LongForm.html">kinMatrix2LongForm()</a></code> </p>
        </td>
        <td><p>Reformats a kinship matrix into a long-format table.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeDuplicates.html">removeDuplicates()</a></code> </p>
        </td>
        <td><p>Remove duplicate records from pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeUninformativeFounders.html">removeUninformativeFounders()</a></code> </p>
        </td>
        <td><p>Remove uninformative founders.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeUnknownAnimals.html">removeUnknownAnimals()</a></code> </p>
        </td>
        <td><p>removeUnknownAnimals Removes unknown animals added to pedigree that serve as
placeholders for unknown parents.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runGenekeepr.html">runGeneKeepR()</a></code> </p>
        </td>
        <td><p>Allows running <code>shiny</code> application with
<code>nprcgenekeepr::runGeneKeepR()</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="trimPedigree.html">trimPedigree()</a></code> </p>
        </td>
        <td><p>Trim pedigree to ancestors of provided group by removing uninformative
individuals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summary.html">summary(<i>&lt;nprcgenekeeprErr&gt;</i>)</a></code> <code><a href="summary.html">summary(<i>&lt;nprcgenekeeprGV&gt;</i>)</a></code> </p>
        </td>
        <td><p>summary.nprcgenekeeprErr Summary function for class nprcgenekeeprErr</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-exposed-functions" class="hasAnchor"><a href="#section-all-exposed-functions" class="anchor"></a>All exposed functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="addAnimalsWithNoRelative.html">addAnimalsWithNoRelative()</a></code> </p>
        </td>
        <td><p>Adds an NA value for all animals without a relative</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addBackSecondParents.html">addBackSecondParents()</a></code> </p>
        </td>
        <td><p>Add back single parents trimmed pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addGenotype.html">addGenotype()</a></code> </p>
        </td>
        <td><p>Add genotype data to pedigree file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addGroupOfUnusedAnimals.html">addGroupOfUnusedAnimals()</a></code> </p>
        </td>
        <td><p>addGroupOfUnusedAnimals adds a group to the saved groups if needed</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addIdRecords.html">addIdRecords()</a></code> </p>
        </td>
        <td><p>addIdRecords Adds Ego records added having NAs for parent IDs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addParents.html">addParents()</a></code> </p>
        </td>
        <td><p>Add parents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addSexAndAgeToGroup.html">addSexAndAgeToGroup()</a></code> </p>
        </td>
        <td><p>Forms a dataframe with Id, Sex, and current Age given a list of Ids and a
pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addUIds.html">addUIds()</a></code> </p>
        </td>
        <td><p>Eliminates partial parentage situations by adding unique placeholder
IDs for the unknown parent.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="agePyramidPlot.html">agePyramidPlot()</a></code> </p>
        </td>
        <td><p>Form age pyramid plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="alleleFreq.html">alleleFreq()</a></code> </p>
        </td>
        <td><p>Calculates the count of each allele in the provided vector.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="allTrueNoNA.html">allTrueNoNA()</a></code> </p>
        </td>
        <td><p>Returns <code>TRUE</code> if every member of the vector is <code>TRUE</code>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assignAlleles.html">assignAlleles()</a></code> </p>
        </td>
        <td><p>Assign parent alleles randomly</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcA.html">calcA()</a></code> </p>
        </td>
        <td><p>Calculates <code>a</code>, the number of an individual's alleles that are rare in
each simulation.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcAge.html">calcAge()</a></code> </p>
        </td>
        <td><p>Calculate animal ages.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcFE.html">calcFE()</a></code> </p>
        </td>
        <td><p>Calculates founder Equivalents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcFEFG.html">calcFEFG()</a></code> </p>
        </td>
        <td><p>Calculates Founder Equivalents and Founder Genome Equivalents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcFG.html">calcFG()</a></code> </p>
        </td>
        <td><p>Calculates Founder Genome Equivalents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcGU.html">calcGU()</a></code> </p>
        </td>
        <td><p>Calculates genome uniqueness for each ID that is part of the population.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcRetention.html">calcRetention()</a></code> </p>
        </td>
        <td><p>Calculates Allelic Retention</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calculateSexRatio.html">calculateSexRatio()</a></code> </p>
        </td>
        <td><p>Calculates the sex ratio (number of non-males / number of males) given
animal Ids and their pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="checkChangedColAndErrorLst.html">checkChangedColAndErrorLst()</a></code> </p>
        </td>
        <td><p>checkChangedColAndErrorLst examines errorLst for errors and
errorLst$changeCols non-empty fields</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="checkChangedColsLst.html">checkChangedColsLst()</a></code> </p>
        </td>
        <td><p>checkChangedColsLst examines list for non-empty fields</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="checkErrorLst.html">checkErrorLst()</a></code> </p>
        </td>
        <td><p>checkErrorLst examines list for non-empty fields</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="checkGenotypeFile.html">checkGenotypeFile()</a></code> </p>
        </td>
        <td><p>Check genotype file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="checkParentAge.html">checkParentAge()</a></code> </p>
        </td>
        <td><p>Check parent ages to be at least <code>minParentAge</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="checkRequiredCols.html">checkRequiredCols()</a></code> </p>
        </td>
        <td><p>Examines column names, <code>cols</code> for required column names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="chooseAlleles.html">chooseAlleles()</a></code> </p>
        </td>
        <td><p>Combines two vectors of alleles by randomly selecting one allele
or the other at each position.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="chooseAllelesChar.html">chooseAllelesChar()</a></code> </p>
        </td>
        <td><p>Combines two vectors of alleles when alleles are character vectors.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="chooseDate.html">chooseDate()</a></code> </p>
        </td>
        <td><p>Choose date based on <code>earlier</code> flag.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="colChange.html">colChange()</a></code> </p>
        </td>
        <td><p>colChange internal function to describe column names transformation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertAncestry.html">convertAncestry()</a></code> </p>
        </td>
        <td><p>Converts the ancestry information to a standardized code</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertDate.html">convertDate()</a></code> </p>
        </td>
        <td><p>Converts date columns formatted as characters to be of type datetime</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertFromCenter.html">convertFromCenter()</a></code> </p>
        </td>
        <td><p>Converts the fromCenter information to a standardized code</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertRelationships.html">convertRelationships()</a></code> </p>
        </td>
        <td><p>Converts pairwise kinship values to a relationship category descriptor.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertSexCodes.html">convertSexCodes()</a></code> </p>
        </td>
        <td><p>Converts sex indicator for an individual to a standardized codes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertStatusCodes.html">convertStatusCodes()</a></code> </p>
        </td>
        <td><p>Converts status indicators to a Standardized code</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="correctParentSex.html">correctParentSex()</a></code> </p>
        </td>
        <td><p>Sets sex for animals listed as either a sire or dam.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="countFirstOrder.html">countFirstOrder()</a></code> </p>
        </td>
        <td><p>Count first-order relatives.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="countLoops.html">countLoops()</a></code> </p>
        </td>
        <td><p>Count the number of loops in a pedigree tree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createExampleFiles.html">createExampleFiles()</a></code> </p>
        </td>
        <td><p>Creates a folder with CSV files containing example pedigrees and ID lists
used to demonstrate the package.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createPedOne.html">createPedOne()</a></code> </p>
        </td>
        <td><p>createPedOne makes the pedOne data object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createPedSix.html">createPedSix()</a></code> </p>
        </td>
        <td><p>createPedSix makes the pedSix data object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createPedTree.html">createPedTree()</a></code> </p>
        </td>
        <td><p>Create a pedigree tree (PedTree).</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="create_wkbk.html">create_wkbk()</a></code> </p>
        </td>
        <td><p>Creates an Excel workbook with worksheets.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dataframe2string.html">dataframe2string()</a></code> </p>
        </td>
        <td><p>dataframe2string converts a data.frame object to a character vector</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fillBins.html">fillBins()</a></code> </p>
        </td>
        <td><p>fillBins Fill bins represented by list of two lists <code>males</code> and
<code>females</code>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fillGroupMembers.html">fillGroupMembers()</a></code> </p>
        </td>
        <td><p>Forms and fills list of animals groups based on provided constraints</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fillGroupMembersWithSexRatio.html">fillGroupMembersWithSexRatio()</a></code> </p>
        </td>
        <td><p>Forms breeding group(s) with an effort to match a specified sex ratio</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterAge.html">filterAge()</a></code> </p>
        </td>
        <td><p>Removes kinship values where an animal is less than the minAge</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterKinMatrix.html">filterKinMatrix()</a></code> </p>
        </td>
        <td><p>Filters a kinship matrix to include only the egos listed in 'ids'</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterPairs.html">filterPairs()</a></code> </p>
        </td>
        <td><p>Filters kinship values from a long-format kinship table based on the sexes
 of the two animals involved.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterReport.html">filterReport()</a></code> </p>
        </td>
        <td><p>Filters a genetic value report down to only the specified animals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterThreshold.html">filterThreshold()</a></code> </p>
        </td>
        <td><p>Filters kinship to remove rows with kinship values less than the specified
threshold</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findGeneration.html">findGeneration()</a></code> </p>
        </td>
        <td><p>Determines the generation number for each id.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findLoops.html">findLoops()</a></code> </p>
        </td>
        <td><p>Find loops in a pedigree tree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findOffspring.html">findOffspring()</a></code> </p>
        </td>
        <td><p>Finds the number of total offspring for each animal in the provided pedigree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findPedigreeNumber.html">findPedigreeNumber()</a></code> </p>
        </td>
        <td><p>Determines the generation number for each id.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fixColumnNames.html">fixColumnNames()</a></code> </p>
        </td>
        <td><p>fixColumnNames changes original column names and into standardized names.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fixGenotypeCols.html">fixGenotypeCols()</a></code> </p>
        </td>
        <td><p>Reformat names of observed genotype columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneDrop.html">geneDrop()</a></code> </p>
        </td>
        <td><p>Gene drop simulation based on the provided pedigree information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_and_or_list.html">get_and_or_list()</a></code> </p>
        </td>
        <td><p>Returns a one element character string with correct punctuation for
a list made up of the elements of the character vector argument.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="get_elapsed_time_str.html">get_elapsed_time_str()</a></code> </p>
        </td>
        <td><p>Returns the elapsed time since start_time.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getAncestors.html">getAncestors()</a></code> </p>
        </td>
        <td><p>Recursively create a character vector of ancestors for an individual ID.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getAnimalsWithHighKinship.html">getAnimalsWithHighKinship()</a></code> </p>
        </td>
        <td><p>Forms a list of animal Ids and animals related to them</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getChangedColsTab.html">getChangedColsTab()</a></code> </p>
        </td>
        <td><p>getChangedColsTab skeleton of list of errors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getConfigFileName.html">getConfigFileName()</a></code> </p>
        </td>
        <td><p>getConfigFileName returns the configuration file name appropriate for
the system.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getCurrentAge.html">getCurrentAge()</a></code> </p>
        </td>
        <td><p>Age in years using the provided birthdate.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getDateColNames.html">getDateColNames()</a></code> </p>
        </td>
        <td><p>Vector of date column names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getDatedFilename.html">getDatedFilename()</a></code> </p>
        </td>
        <td><p>Returns a character vector with an file name having the date prepended.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getDateErrorsAndConvertDatesInPed.html">getDateErrorsAndConvertDatesInPed()</a></code> </p>
        </td>
        <td><p>Converts columns of dates in text form to <code>Date</code> object columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getDemographics.html">getDemographics()</a></code> </p>
        </td>
        <td><p>Get demographic data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getEmptyErrorLst.html">getEmptyErrorLst()</a></code> </p>
        </td>
        <td><p>Creates a empty errorLst object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getErrorTab.html">getErrorTab()</a></code> </p>
        </td>
        <td><p>getErrorTab skeleton of list of errors</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getFocalAnimalPed.html">getFocalAnimalPed()</a></code> </p>
        </td>
        <td><p>Get pedigree based on list of focal animals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGenoDefinedParentGenotypes.html">getGenoDefinedParentGenotypes()</a></code> </p>
        </td>
        <td><p>Assigns parental genotype contributions to an IDs genotype by attributing
alleles to sire or dam</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGenotypes.html">getGenotypes()</a></code> </p>
        </td>
        <td><p>Get genotypes from file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGVGenotype.html">getGVGenotype()</a></code> </p>
        </td>
        <td><p>Get Genetic Value Genotype data structure for reportGV function.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGVPopulation.html">getGVPopulation()</a></code> </p>
        </td>
        <td><p>Get the population of interest for the Genetic Value analysis.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getIdsWithOneParent.html">getIdsWithOneParent()</a></code> </p>
        </td>
        <td><p>getIdsWithOneParent extracts IDs of animals pedigree without either a
sire or a dam</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getIncludeColumns.html">getIncludeColumns()</a></code> </p>
        </td>
        <td><p>Get the superset of columns that can be in a pedigree file.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getIndianOriginStatus.html">getIndianOriginStatus()</a></code> </p>
        </td>
        <td><p>Get Indian-origin status of group</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getLkDirectAncestors.html">getLkDirectAncestors()</a></code> </p>
        </td>
        <td><p>Get the direct ancestors of selected animals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getLkDirectRelatives.html">getLkDirectRelatives()</a></code> </p>
        </td>
        <td><p>Get the direct ancestors of selected animals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getLogo.html">getLogo()</a></code> </p>
        </td>
        <td><p>Get Logo file name</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getMaxAx.html">getMaxAx()</a></code> </p>
        </td>
        <td><p>Get the maximum of the absolute values of the negative (males) and positive
(female) animal counts.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getMinParentAge.html">getMinParentAge()</a></code> </p>
        </td>
        <td><p>Get minimum parent age.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getOffspring.html">getOffspring()</a></code> </p>
        </td>
        <td><p>Get offspring to corresponding animal IDs provided</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getParamDef.html">getParamDef()</a></code> </p>
        </td>
        <td><p>Get parameter definitions from tokens found in configuration file.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getParents.html">getParents()</a></code> </p>
        </td>
        <td><p>Get parents to corresponding animal IDs provided</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPedigree.html">getPedigree()</a></code> </p>
        </td>
        <td><p>Get pedigree from file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPedMaxAge.html">getPedMaxAge()</a></code> </p>
        </td>
        <td><p>Get the maximum age of live animals in the pedigree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPossibleCols.html">getPossibleCols()</a></code> </p>
        </td>
        <td><p>Get possible column names for a studbook.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPotentialSires.html">getPotentialSires()</a></code> </p>
        </td>
        <td><p>Provides list of potential sires</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getProbandPedigree.html">getProbandPedigree()</a></code> </p>
        </td>
        <td><p>Gets pedigree to ancestors of provided group leaving uninformative ancestors.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getProductionStatus.html">getProductionStatus()</a></code> </p>
        </td>
        <td><p>Get production status of group</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getProportionLow.html">getProportionLow()</a></code> </p>
        </td>
        <td><p>Get proportion of Low genetic value animals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPyramidAgeDist.html">getPyramidAgeDist()</a></code> </p>
        </td>
        <td><p>Get the age distribution for the pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPyramidPlot.html">getPyramidPlot()</a></code> </p>
        </td>
        <td><p>Creates a pyramid plot of the pedigree provided.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getRecordStatusIndex.html">getRecordStatusIndex()</a></code> </p>
        </td>
        <td><p>Returns record numbers with selected <code>recordStatus</code>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getRequiredCols.html">getRequiredCols()</a></code> </p>
        </td>
        <td><p>Get required column names for a studbook.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getSiteInfo.html">getSiteInfo()</a></code> </p>
        </td>
        <td><p>Get site information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getTokenList.html">getTokenList()</a></code> </p>
        </td>
        <td><p>Gets tokens from character vector of lines</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getVersion.html">getVersion()</a></code> </p>
        </td>
        <td><p>getVersion Get the version number of nprcgenekeepr</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="groupAddAssign.html">groupAddAssign()</a></code> </p>
        </td>
        <td><p>Add animals to an existing breeding group or forms groups:</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="groupMembersReturn.html">groupMembersReturn()</a></code> </p>
        </td>
        <td><p>Forms return list of groupAddAssign function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hasGenotype.html">hasGenotype()</a></code> </p>
        </td>
        <td><p>Check for genotype data in dataframe</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hasBothParents.html">hasBothParents()</a></code> </p>
        </td>
        <td><p>hasBothParents checks to see if both parents are identified.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="headerDisplayNames.html">headerDisplayNames()</a></code> </p>
        </td>
        <td><p>Convert internal column names to display or header names.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="initializeHaremGroups.html">initializeHaremGroups()</a></code> </p>
        </td>
        <td><p>Make the initial groupMembers animal list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="insertChangedColsTab.html">insertChangedColsTab()</a></code> </p>
        </td>
        <td><p>insertChangedColsTab insert a list of changed columns found by
<code>qcStudbook</code> in the pedigree file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="insertErrorTab.html">insertErrorTab()</a></code> </p>
        </td>
        <td><p>insertErrorTab insert a list of errors found by <code>qcStudbook</code> in the
pedigree file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="insertSeparators.html">insertSeparators()</a></code> </p>
        </td>
        <td><p>insertSeparators inserts the character "-" between year and month and
between month and day portions of a date string in %Y%m%d format.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="isEmpty.html">isEmpty()</a></code> </p>
        </td>
        <td><p>Is vector empty or all NA values.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="is_valid_date_str.html">is_valid_date_str()</a></code> </p>
        </td>
        <td><p>Returns TRUE if the string is a valid date.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="kinship.html">kinship()</a></code> </p>
        </td>
        <td><p>Generates a kinship matrix.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeAvailable.html">makeAvailable()</a></code> </p>
        </td>
        <td><p>Convenience function to make the initial available animal list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeCEPH.html">makeCEPH()</a></code> </p>
        </td>
        <td><p>Make a CEPH-style pedigree for each id</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeExamplePedigreeFile.html">makeExamplePedigreeFile()</a></code> </p>
        </td>
        <td><p>Write copy of nprcgenekeepr::examplePedigree into a file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeGroupMembers.html">makeGroupMembers()</a></code> </p>
        </td>
        <td><p>Convenience function to make the initial groupMembers animal list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeGrpNum.html">makeGrpNum()</a></code> </p>
        </td>
        <td><p>Convenience function to make the initial grpNum list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeRelationClassesTable.html">makeRelationClassesTable()</a></code> </p>
        </td>
        <td><p>Make relation classes table from <code>kin</code> dataframe.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeRoundUp.html">makeRoundUp()</a></code> </p>
        </td>
        <td><p>Round up the provided integer vector <code>int</code> according to the
<code>modulus</code>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makesLoop.html">makesLoop()</a></code> </p>
        </td>
        <td><p><code>makesLoop</code> tests for a common ancestor.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mapIdsToObfuscated.html">mapIdsToObfuscated()</a></code> </p>
        </td>
        <td><p>Map IDs to Obfuscated IDs</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="meanKinship.html">meanKinship()</a></code> </p>
        </td>
        <td><p>Calculates the mean kinship for each animal in a kinship matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="offspringCounts.html">offspringCounts()</a></code> </p>
        </td>
        <td><p>Finds the total number of offspring for each animal in the pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="obfuscateDate.html">obfuscateDate()</a></code> </p>
        </td>
        <td><p>obfucateDate adds a random number of days bounded by plus and minus max delta</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="obfuscateId.html">obfuscateId()</a></code> </p>
        </td>
        <td><p>obfucateId creates a vector of ID aliases of specified length</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="obfuscatePed.html">obfuscatePed()</a></code> </p>
        </td>
        <td><p>obfuscatePed takes a pedigree object and creates aliases for all IDs and
adjusts all date within a specified amount.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="orderReport.html">orderReport()</a></code> </p>
        </td>
        <td><p>Order the results of the genetic value analysis for use in a report.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="print.html">print(<i>&lt;summary.nprcgenekeeprErr&gt;</i>)</a></code> <code><a href="print.html">print(<i>&lt;summary.nprcgenekeeprGV&gt;</i>)</a></code> </p>
        </td>
        <td><p>print.summary.nprcgenekeepr print.summary.nprcgenekeeprGV</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qcStudbook.html">qcStudbook()</a></code> </p>
        </td>
        <td><p>Quality Control for the Studbook or pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rankSubjects.html">rankSubjects()</a></code> </p>
        </td>
        <td><p>Ranks animals based on genetic value.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rbindFill.html">rbindFill()</a></code> </p>
        </td>
        <td><p>Append the rows of one dataframe to another.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeGroupIfNoAvailableAnimals.html">removeGroupIfNoAvailableAnimals()</a></code> </p>
        </td>
        <td><p>Remove group numbers when all available animals have been used</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeSelectedAnimalFromAvailableAnimals.html">removeSelectedAnimalFromAvailableAnimals()</a></code> </p>
        </td>
        <td><p>Updates list of available animals by removing the selected animal</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="kinMatrix2LongForm.html">kinMatrix2LongForm()</a></code> </p>
        </td>
        <td><p>Reformats a kinship matrix into a long-format table.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="readExcelPOSIXToCharacter.html">readExcelPOSIXToCharacter()</a></code> </p>
        </td>
        <td><p>Read in Excel file and convert POSIX dates to character</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeDuplicates.html">removeDuplicates()</a></code> </p>
        </td>
        <td><p>Remove duplicate records from pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeEarlyDates.html">removeEarlyDates()</a></code> </p>
        </td>
        <td><p>removeEarlyDates removes dates before a specified year</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removePotentialSires.html">removePotentialSires()</a></code> </p>
        </td>
        <td><p>Removes potential sires from list of Ids</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeUninformativeFounders.html">removeUninformativeFounders()</a></code> </p>
        </td>
        <td><p>Remove uninformative founders.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeUnknownAnimals.html">removeUnknownAnimals()</a></code> </p>
        </td>
        <td><p>removeUnknownAnimals Removes unknown animals added to pedigree that serve as
placeholders for unknown parents.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reportGV.html">reportGV()</a></code> </p>
        </td>
        <td><p>Generates a genetic value report for a provided pedigree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="resetGroup.html">resetGroup()</a></code> </p>
        </td>
        <td><p>Update or add the "group" field of a Pedigree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runGenekeepr.html">runGeneKeepR()</a></code> </p>
        </td>
        <td><p>Allows running <code>shiny</code> application with
<code>nprcgenekeepr::runGeneKeepR()</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="saveDataframesAsFiles.html">saveDataframesAsFiles()</a></code> </p>
        </td>
        <td><p>Write copy of dataframes to either CSV or Excel file.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="setPopulation.html">setPopulation()</a></code> </p>
        </td>
        <td><p>Population designation function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getSexRatioWithAdditions.html">getSexRatioWithAdditions()</a></code> </p>
        </td>
        <td><p>getSexRatioWithAdditions returns the sex ratio of a group.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="setExit.html">setExit()</a></code> </p>
        </td>
        <td><p>Sets the exit date, if there is no exit column in the table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="set_seed.html">set_seed()</a></code> </p>
        </td>
        <td><p>Work around for unit tests using sample() among various versions of R</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="str_detect_fixed_all.html">str_detect_fixed_all()</a></code> </p>
        </td>
        <td><p>Returns a logical vector with results of stri_detect() for each pattern in
second parameters character vector.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summary.html">summary(<i>&lt;nprcgenekeeprErr&gt;</i>)</a></code> <code><a href="summary.html">summary(<i>&lt;nprcgenekeeprGV&gt;</i>)</a></code> </p>
        </td>
        <td><p>summary.nprcgenekeeprErr Summary function for class nprcgenekeeprErr</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="toCharacter.html">toCharacter()</a></code> </p>
        </td>
        <td><p>Force dataframe columns to character</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="trimPedigree.html">trimPedigree()</a></code> </p>
        </td>
        <td><p>Trim pedigree to ancestors of provided group by removing uninformative
individuals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="unknown2NA.html">unknown2NA()</a></code> </p>
        </td>
        <td><p>Removing IDs having "UNKNOWN" regardless of case</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="withinIntegerRange.html">withinIntegerRange()</a></code> </p>
        </td>
        <td><p>Get integer within a range</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Michael Raboin, Terry Therneau, Amanda Vinson, R. Mark Sharp, Southwest National Primate Research Center NIH grant P51 RR13986, Oregon National Primate Research Center grant P51 OD011092.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


